packages/hardhat-core/src/builtin-tasks/node.ts

```diff
@@ -66,6 +47,110 @@ Private Key: ${privateKey}
   }
 }
 
+subtask(TASK_NODE_GET_PROVIDER).setAction(
+  async (
+    {
+      forkBlockNumber: forkBlockNumberParam,
+      forkUrl: forkUrlParam,
+    }: {
+      forkBlockNumber?: number;
+      forkUrl?: string;
+    },
+    { config, network }
+  ): Promise<EthereumProvider> => {
+    // enable logging for in-memory hardhat network provider
+    await network.provider.request({
+      method: "hardhat_setLoggingEnabled",
+      params: [true],
+    });
+
+    assertHardhatInvariant(
+      network.name === HARDHAT_NETWORK_NAME,
+      "Hardhat node can only be used with hardhat network"
+    );
+
+    const networkConfig = config.networks[HARDHAT_NETWORK_NAME];
+
+    const forkUrl = forkUrlParam ?? networkConfig.forking?.url;
+    const forkBlockNumber =
+      forkBlockNumberParam ?? networkConfig.forking?.blockNumber;
+
+    // we use the hardhat_reset RPC method to enable the fork
+    if (forkUrl !== undefined) {
+      await network.provider.request({
+        method: "hardhat_reset",
+        params: [
+          {
+            forking: {
+              jsonRpcUrl: forkUrl,
+              blockNumber: forkBlockNumber,
+            },
+          },
+        ],
+      });
+    } else if (forkBlockNumber !== undefined) {
+      // we throw an error if the user specified a forkBlockNumber but not a
+      // forkUrl
+      throw new HardhatError(
+        ERRORS.BUILTIN_TASKS.NODE_FORK_BLOCK_NUMBER_WITHOUT_URL
+      );
+    }
+
+    return network.provider;
+  }
+);
+
+subtask(TASK_NODE_CREATE_SERVER)
+  .addParam("hostname", undefined, undefined, types.string)
+  .addParam("port", undefined, undefined, types.int)
+  .addParam("provider", undefined, undefined, types.any)
+  .setAction(
+    async ({
+      hostname,
+      port,
+      provider,
+    }: {
+      hostname: string;
+      port: number;
+      provider: EthereumProvider;
```

Comment:
Maybe `EIP1193Provider`? It's an easier to implement interface.

Reply:
Forget what I just said. That would be annoying for people overriding these tasks.